Simplify effect parameters

Paul-Licameli opened the pull request:
Resolves: #2679 

Still a very drafty draft, based on the waiting PR #2676 

I invite the start of reviews now.  The fourth commit is big.

*(short description of the changes and the motivation to make the changes)*

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

pietro68 commented on src/effects/FindClipping.cpp:
```diff
@@ -96,8 +96,8 @@ bool EffectFindClipping::GetAutomationParameters(CommandParameters & parms) cons
 {
-   ReadAndVerifyInt(Start);
-   ReadAndVerifyInt(Stop);
+   ReadParam(Start);
```
Now that `ReadParam` resolves types at compile time, we have no more need for the many `ReadAndVerifyInt`, `ReadAndVerifyBool `etc.  - however, in the process we lost the verb "verify" in the function name , which is what ReadParam actually does. So, could we rename `ReadParam` to `ReadAndVerify`?

Paul-Licameli replied:
The commit "Remove unused macros and the cache in EffectParameter" ultimately eliminates these macros.  This is just one simplifying step in my transformation.

pietro68 replied:
ok

pietro68 commented on src/effects/Effect.h:
```diff
@@ -133,6 +138,11 @@ class AUDACITY_DLL_API Effect /* not final */ : public wxEvtHandler,
+   // use the functions of EffectParameterMethods.  By default, this function
+   // defines an empty list of parameters.
+   virtual const EffectParameterMethods &Parameters() const;
+
```
Proposal: change the name of this method to GetEffectParameterMethods

pietro68 commented on src/effects/Equalization.h:
```diff
@@ -274,6 +276,9 @@ class EffectEqualization : public Effect,
    wxBoxSizer *szrM;
 #endif
 
+   CapturedParameters<EffectEqualization> mParameters;
```
What is this member for? we already have the static capture `parameters` in `EffectEqualization::Parameters`

Paul-Licameli replied:
Good find.  That was left over from an earlier version of this pull request, in which the parameters object still needed to capture `this` and could not be a static.  I left the same in TruncSilence.h.

pietro68 commented on src/ShuttleAutomation.h:
```diff
@@ -11,85 +11,293 @@
+template <typename EffectType, const auto &...Parameters>
+class CapturedParameters : public EffectParameterMethods {
+public:
+   static constexpr auto *fetcher = &EffectType::FetchParameters;
```
On this line, using msvc2019 I get:
`error C3535: cannot deduce type for 'auto *' from 'int'`

pietro68 commented on src/Shuttle.h:
```diff
@@ -56,34 +107,51 @@ class CommandParameters;
+   virtual void DefineEnum( Arg<int> var, const wxChar * key, int vdefault,
       const EnumValueSymbol strings[], size_t nStrings );
 };
 
```
On these two I get this warnings on msvc2019:

`warning C4910: 'SettingsVisitorBase<false>': '__declspec(dllexport)' and 'extern' are incompatible on an explicit instantiation`

Not sure how serious this warning is though: I can not get to the link stage because I get a compile error (which I noted elsewhere)

pietro68 commented on src/ShuttleAutomation.h:
```diff
@@ -11,85 +11,293 @@
+   }
+   // More specific overload for enumeration parameters
+   template< typename Member >
+   static bool SetOne(Effect &effect, Params &structure,
```
as above

Paul-Licameli commented on src/ShuttleAutomation.h:
```diff
@@ -11,85 +11,293 @@
+   }
+
+   template< typename Member, typename Type, typename Value >
+   static bool SetOne(Effect &effect, Params &structure,
```
Good catch

Paul-Licameli commented:
I have updated the branch, fixing the unused parameter Pietro pointed out.

I also did a self-review of "Use CapturedParameters to simplify some effects..." and fixed some details, where there are post-set functions used in the construction of parameters.

pietro68 reviewed (approved):
I think I understood the code, save from a few bits in the CapturedParameters section - but that code has changed a bit since this review was opened, so I am going to study that again.
